[FLINK-36808][table-planner] Fix LookupJoin bug when used with filterable table source

morazow opened the pull request:
## What is the purpose of the change

- Fix bug in LookupJoin that occurs when used together with filterable (that support filter pushdown) table and union all

### Brief description of the bug

Given a union all query:
- that combines results of the two lookup joins
- that these lookup joins have different filter queries

```
SELECT
     s.id,
     s.name,
     s.txn_time,
     d.status
FROM `stream` AS `s` INNER JOIN `dim` FOR SYSTEM_TIME AS OF `s`.`txn_time` AS `d`
ON
     `s`.`id` = `d`.`id`
WHERE
     `d`.`status` = 'OK' 
UNION ALL
SELECT
     s.id,
     s.name,
     s.txn_time,
     d.status
FROM `stream` AS `s` INNER JOIN `dim` FOR SYSTEM_TIME AS OF `s`.`txn_time` AS `d`
ON
     `s`.`id` = `d`.`id`
WHERE
     `d`.`status` = 'NOT_EXISTS';
```

In this situation the planner will pushdown the filter condition into the table part of the lookup join, but the structure of the lookup joins stays the same, e.g, they will have same `digests` with different table / temporal table.

This is the problem since the Calcite Volcano optimizer will register them equivalent because it does so [using the digest of the relation nodes](https://github.com/apache/calcite/blob/main/core/src/main/java/org/apache/calcite/plan/volcano/VolcanoPlanner.java#L1304-L1328).

This introduces the bug because when optimizing the Union, both parts of the query will be treated same (even though we have different where clauses) and the found cheapest plan will be same for both lookup joins.

You can also see the effect if you put non-existing filter first the result will be empty, because the first lookupjoin is also used for the second part of the union.

### Alternative Solutions

The better solution would be to improve the LookupJoin expression to also include the filter condition into the table name. For example,
```
LookupJoin(table=[default_catalog.default_database.dim, filter=[<PUSHED-DOWN-FILTERS>]], joinType=[InnerJoin], lookup=[id=id], select=[id, name, txn_time, id], upsertKey=[[0]])
```

But this would require many refactoring, and mainly in the tests.

In this PR, I have opted for adding another `filter` with pushed down filter conditions if the LookupJoin contains a table with filter pushdowns.

## Brief change log

- Add unit test to reproduce the test
- Add (one alternative) fix to resolve the bug

## Verifying this change

The change adds test case that reproduces the bug that can be verified by the fixes.

## Does this pull request potentially affect one of the following parts:

- Dependencies (does it add or upgrade a dependency): no
- The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
- The serializers: no
- The runtime per-record code paths (performance sensitive): no
- Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
- The S3 file system connector: no 

## Documentation

- Does this pull request introduce a new feature? no
- If yes, how is the feature documented? not applicable

morazow committed:
[FLINK-36808][table-planner] Fix LookupJoin bug when used with filterable table source

davidradl commented:
nit: the Jira number has an extra `0` at the end in the PR title

davidradl commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/common/CommonPhysicalLookupJoin.scala:
```diff
@@ -21,16 +21,18 @@ import org.apache.flink.table.api.{TableConfig, TableException}
 import org.apache.flink.table.planner.plan.nodes.physical.stream.StreamPhysicalRel
 import org.apache.flink.table.planner.plan.schema.{IntermediateRelTable, LegacyTableSourceTable, TableSourceTable}
-import org.apache.flink.table.planner.plan.utils.{ChangelogPlanUtils, ExpressionFormat, InputRefVisitor, JoinTypeUtil, LookupJoinUtil, RelExplainUtil, TemporalJoinUtil}
+import org.apache.flink.table.planner.plan.utils._
```
nit: normal practise is to list all of the import packages.

morazow commented:
Thanks @davidradl, addressed your findings, please have another look 🤝 

> the filtered pushed down joins cause the Volcano planner to uniquely identify each look up join.

Yes, but based on the getDigests method, which results in a string like below:

```
LookupJoin(table=[default_catalog.default_database.dim], joinType=[InnerJoin], lookup=[id=id], select=[id, name, txn_time, id], upsertKey=[[0]])
```

Here we have the table name, but no information on the pushed down filter conditions. For this issue to happen, the lookup table's should be same (since it is in the digest), so we cannot test if one part supports and other part doesn't test case. 

Additionally, if the source doesn't support the filter pushdown it will be represented in the digest with the `where` keyword. This will not also reproduce the bug because the digests are different.

xuyangzhong commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/common/CommonPhysicalLookupJoin.scala:
```diff
@@ -209,6 +215,20 @@ abstract class CommonPhysicalLookupJoin(
   }
 
+  private def getTableFilterString(t: TableSourceTable): String = {
+    val filterOpt = t.abilitySpecs.collectFirst { case spec: FilterPushDownSpec => spec }
```
I'm wondering should `PartitionPushDownSpec` also need to be added into this part.

morazow replied:
Added 👍

xuyangzhong commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/common/CommonPhysicalLookupJoin.scala:
```diff
@@ -190,13 +191,18 @@ abstract class CommonPhysicalLookupJoin(
       .item("joinType", JoinTypeUtil.getFlinkJoinType(joinType))
       .item("lookup", lookupKeys)
       .itemIf("where", whereString, whereString.nonEmpty)
+      .itemIf("filter", filterPushdownString, filterPushdownString.nonEmpty)
```
nit what about `filterPushedDown`?

xuyangzhong commented on flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/UnionLookupJoinITCase.scala:
```diff
@@ -0,0 +1,96 @@
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.CsvSource
+
+class UnionLookupJoinITCase extends StreamingTestBase {
```
Could you please move this test to `LookupJoinITCase`?

xuyangzhong reviewed (commented):
Thanks for digging into this bug and driving this fix. I just left some comments.

morazow committed:
[FLINK-36808][table-planner] Address PR reviews suggestions

morazow commented:
Hello @xuyangzhong ,

Thanks for the review! I have addressed your suggestions, please have a look. I put them to separate commit for now, later I'll squash it 👍

xuyangzhong commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/common/CommonPhysicalLookupJoin.scala:
```diff
@@ -190,13 +191,18 @@ abstract class CommonPhysicalLookupJoin(
 
     super
       .explainTerms(pw)
       .item("table", tableIdentifier.asSummaryString())
```
What about reusing this item `table` with:
```
val tableDesc: String = temporalTable match {
  case t: TableSourceTable => t.getQualifiedName.asScala.mkString(", ")
  case t: LegacyTableSourceTable[_] => t.tableIdentifier.asSummaryString()
}

super
  .explainTerms(pw)
  .item("table", tableDesc)
  ...
```

Although the plans for some tests may change, this will not actually affect compatibility, and it can be aligned with the digest in `CommonPhysicalTableSourceScan`.

xuyangzhong reviewed (commented):
Thanks for the update. Please forgive me for missing your message... I have one comment as well.
